### PR TITLE
WIP: Upgrade kubebuilder to v2.3.1

### PR DESCRIPTION
**Pull request by jackatbancast**
This upgrade involves upgrades to go.mod and newer versions of kubernetes libs and controller-{runtime,gen}.

Most of the remaining work is copying the controller code, regenerating the webhooks and re-adding the tests.

### CODE REVIEW

**Comment by jackatbancast** [comment]
This will not work in kubernetes v1.18.x until https://github.com/kubernetes/kubernetes/pull/91419 is merged, or kubebuilder is fixed.

File: Dockerfile [hunk 1]
```diff
@@ -12,7 +14,8 @@ RUN set -x \
       && mv pkg/linux_amd64/envconsul ../bin
 
 COPY . /go/src/github.com/gocardless/theatre
-RUN make VERSION=$(cat VERSION) build
+
+RUN make -j 3 VERSION=$(cat VERSION) build
```
Comment (lawrencejones) [inline Dockerfile:18]: This usually isn't a good idea, as Go parallelises everything internally anyway. I'd recommend removing the -j for that reason.

File: Makefile [hunk 2]
```diff
@@ -1,10 +1,16 @@
-PROG=bin/rbac-manager bin/workloads-manager bin/vault-manager bin/theatre-envconsul bin/theatre-consoles
```
Comment (lawrencejones) [inline Makefile:1]: There's loads of changes in this file that are totally orthogonal to kube-builder v2. Can we strip this and everything else like it out of this PR? Should make it much more manageable

**Comment by rnaveiras** [comment]
Closed this PR in favour of small PR's
https://github.com/gocardless/theatre/milestone/2?closed=1
